crates/ruff_python_semantic/src/red_knot/types/infer.rs

```diff
@@ -0,0 +1,980 @@
+        });
+
+        // Skip over the function or type params child scope.
+        let (_, scope) = self.children_scopes.next().unwrap();
+
+        assert!(matches!(
+            scope.kind(),
+            ScopeKind::Function | ScopeKind::Annotation
+        ));
+
+        self.definition_tys
+            .insert(Definition::FunctionDef(function_id), function_ty);
+    }
+
+    fn lower_class_definition_statement(&mut self, class: &ast::StmtClassDef) {
+        let ast::StmtClassDef {
+            range: _,
+            name,
+            type_params,
+            decorator_list,
+            arguments,
+            body: _,
+        } = class;
+
+        let class_id = class.scope_ast_id(self.db, self.file_id, self.file_scope_id);
+
+        for decorator in decorator_list {
+            self.lower_decorator(decorator);
+        }
+
+        let bases = arguments
+            .as_deref()
+            .map(|arguments| self.lower_arguments(arguments))
+            .unwrap_or(Vec::new());
+
+        // If the class has type parameters, then the function scope is the first child scope of the type parameter's scope
+        // Otherwise the next scope must be the class definition scope.
+        let (class_body_scope_id, class_body_scope) = if type_params.is_some() {
+            let (type_params_scope, _) = self.children_scopes.next().unwrap();
+            self.index.child_scopes(type_params_scope).next().unwrap()
+        } else {
+            self.children_scopes.next().unwrap()
+        };
+
+        assert_eq!(class_body_scope.kind(), ScopeKind::Class);
+
+        let class_ty = self.class_ty(ClassType {
+            name: Name::new(name),
+            bases,
+            body_scope: class_body_scope_id.to_scope_id(self.db, self.file_id),
+        });
+
+        self.definition_tys
+            .insert(Definition::ClassDef(class_id), class_ty);
+    }
+
+    fn lower_if_statement(&mut self, if_statement: &ast::StmtIf) {
+        let ast::StmtIf {
+            range: _,
+            test,
+            body,
+            elif_else_clauses,
+        } = if_statement;
+
+        self.lower_expression(test);
+        self.lower_body(body);
+
+        for clause in elif_else_clauses {
+            let ast::ElifElseClause {
+                range: _,
+                test,
+                body,
+            } = clause;
+
+            if let Some(test) = &test {
+                self.lower_expression(test);
+            }
+
+            self.lower_body(body);
+        }
+    }
+
+    fn lower_assignment_statement(&mut self, assignment: &ast::StmtAssign) {
+        let ast::StmtAssign {
+            range: _,
+            targets,
+            value,
+        } = assignment;
+
+        let value_ty = self.lower_expression(value);
+
+        for target in targets {
+            self.lower_expression(target);
+        }
+
+        let assign_id = assignment.scope_ast_id(self.db, self.file_id, self.file_scope_id);
+
```

Comment:
There were some important TODO comments in the previous code that are missing here, about handling unpacking assignment, e.g. `x, (y, z) = foo()`. This is going to be a bit tricky to handle, since we will need to construct different `Definition` for all three variables on the LHS, and somehow represent in those Definition the "location" of each var in the unpacking expression; but it can't just be a simple index like for import Definitions.

If we can now have AST IDs for any node, I wonder if both for imports and for assignments it would be better to keep in the Definition the AST ID of the relevant sub-node (i.e. the Name with Store ctx on the LHS for assignment, and the alias for imports) rather than the outer statement.

Reply:
> If we can now have AST IDs for any node, I wonder if both for imports and for assignments it would be better to keep in the Definition the AST ID of the relevant sub-node (i.e. the Name with Store ctx on the LHS for assignment, and the alias for imports) rather than the outer statement.

Possibly. I think I had this in my prototype PR but then run into issues where I needed to go from `AstId` -> `Node` where I would only get the `Alias` which isn't useful (or at least not useful enough to do type inference) and our AST has no mechanism to do upward traversal. This version no longer needs this functionality. So maybe it's okay to just store the specific target / alias.